docs/source-pytorch/advanced/model_parallel/fsdp.rst

```diff
@@ -2,138 +2,444 @@
-parameter sharding with manual wrapping, you can wrap your model using the ``wrap`` function. Internally in Lightning, we enable a context manager around the ``configure_model`` hook to make sure the ``wrap`` parameters are passed correctly.
 
-When not using Fully Sharded, these ``wrap`` calls are a no-op. This means once the changes have been made, there is no need to remove the changes for other strategies.
 
-``wrap`` simply wraps the module with a Fully Sharded Parallel class with the correct parameters from the Lightning context manager.
+We will reuse this Transformer example throughout the guide, optimize speed and memory usage, and compare it to regular DDP training.
 
-Here's an example using that uses ``wrap`` to create your model:
+
+----
+
+
+*********************
+Identify large layers
+*********************
+
+Models that have many large layers like linear layers in LLMs, ViTs, etc. with >100M parameters will benefit the most from FSDP because the memory they consume through parameters, activations and corresponding optimizer states can be evenly split across all GPUs.
+However, one should avoid splitting small layers that have a few thousand parameters because communication overhead would dominate and slow the training down.
+We can specify a list of layer classes in the **wrapping policy** to inform FSDP which parameters it should wrap:
 
 .. code-block:: python
 
-    import torch
-    import torch.nn as nn
-    import lightning.pytorch as pl
-    from lightning.pytorch import Trainer
-    from torch.distributed.fsdp.wrap import wrap
+    # 1. Define a set of layers that FSDP should manage
+    #    Here we are choosing the large encoder and decoder layers
+    policy = {nn.TransformerEncoderLayer, nn.TransformerDecoderLayer}
+
+    # 2. Pass the policy to the FSDPStrategy object
+    strategy = FSDPStrategy(auto_wrap_policy=policy)
+
+    trainer = L.Trainer(..., strategy=strategy)
+
+.. collapse:: Alternative ways to define the policy (Lightning < 2.1)
+
+    The ``auto_wrap_policy`` argument also accepts the old-style function-policies. For example:
+
+    .. code-block:: python
+
+        from functools import partial
+
+        # 1. Import a suiting wrapping policy from PyTorch
+        from torch.distributed.fsdp.wrap import size_based_auto_wrap_policy
+
+        # 2. Configure the policy
+        policy = partial(size_based_auto_wrap_policy, min_num_params=10000)
+
+        # 3. Pass it to the FSDPStrategy object
+        strategy = FSDPStrategy(auto_wrap_policy=policy)
+
+    PyTorch provides several of these functional policies under :mod:`torch.distributed.fsdp.wrap`.
+
+|
+
+Verify that FSDP works with your model by comparing the peak memory usage printed in the CUDA memory summary (see example above) with regular DDP training.
+You should see a decrease in allocated memory and a slight increase in iteration time:
+
+.. list-table::
```

Comment:
Is it the example above and what HW was used?

Reply:
Yes, the exact example as shown in this page was used to get these numbers. Sure, I can add the hardware info, good idea.